File: 프로그래머스/lv0/120861. 캐릭터의 좌표/캐릭터의 좌표.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> keyinput, vector<int> board)
{
    vector<int> answer;
    int idx = 0, x = 0, y = 0, dx = board[0] / 2, dy = board[1] / 2;

    while (1)
    {
        if (idx == keyinput.size())
            break;

        if (keyinput[idx] == "up" && y + 1 <= dy)
            y += 1;
        else if (keyinput[idx] == "down" && y - 1 >= -dy)
            y -= 1;
        else if (keyinput[idx] == "left" && x - 1 >= -dx)
            x -= 1;
        else if (keyinput[idx] == "right" && x + 1 <= dx)
            x += 1;
        idx++;
    }

    answer.push_back(x);
    answer.push_back(y);

    return answer;
}

File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.java
import java.util.Map;

class Solution {

    private final Map<String, int[]> map = Map.of(
            "left", new int[]{-1, 0},
            "right", new int[]{1, 0},
            "up", new int[]{0, 1},
            "down", new int[]{0, -1}
    );

    public int[] solution(String[] keyinput, int[] board) {
        int[] answer = {0, 0};
        int x = board[0] / 2, y = board[1] / 2;

        for (String s : keyinput) {
            int[] move = map.get(s);
            if (Math.abs(answer[0] + move[0]) <= x && Math.abs(answer[1] + move[1]) <= y) {
                answer[0] += move[0];
                answer[1] += move[1];
            }

        }

        return answer;
    }
}